### DISCUSSION TITLE
[Abandoned!] Suggestion for Patch System

### DISCUSSION THREAD

**Discussion by 0lhi** [Wohlstand/TheXTech#355] [Locked]
In order to fix Vanilla Levels that are broken by new features like Multipoints or Multires, a system to patch old levels on-the-fly is required.

I'd like to make a suggestion how this could be implemented by the example of Cliffside.lvl from [Super RMN Bros.](https://rpgmaker.net/games/2015/):

![Cliffside](https://user-images.githubusercontent.com/74732674/179976170-4c65ce9c-fe5f-48c5-aa94-a4abcedb4452.png)

1. Copy the old level inside a folder named ```OLD``` and check it's MD5-Checksum (here: 33e39e089734fa2b679898c40b5d6d77).

2. Add the changes to the level via an Editor (preferably Moondust with the [1.3 Config Pack](https://wohlsoft.ru/config_packs/pack_info.php?pack=SMBX13/)) and save as ```NEW/Cliffside.lvl```.

3. Create the Patch with the following Command*: ```diff -u OLD/Cliffside.lvl NEW/Cliffside.lvl > 33e39e089734fa2b679898c40b5d6d77.patch```
<sub>*Works on Linux, don't know about Windows.</sub>

This will give us:

```diff
--- OLD/Cliffside.lvl	2010-04-12 20:29:18.000000000 +0200
+++ NEW/Cliffside.lvl	2022-07-20 11:40:55.298394121 +0200
@@ -12,10 +12,10 @@
 #FALSE#
 #FALSE#
 "Cavi Cape.mp3"
-0
-0
-0
-0
+-180000
+-180600
+-180000
+-179200
 0
 16291944
 #FALSE#
@@ -24,10 +24,10 @@
[...]
```

Anything before the ```---``` / ```+++``` lines will be ignored. This allows us:

4. To add comments about the changes made:

```diff
#1) Fill empty blocks for Multires.
#2) Add  bgo-61 (Black 32x32) to some pipes and blocks for improved looks.
--- OLD/Cliffside.lvl	2010-04-12 20:29:18.000000000 +0200
+++ NEW/Cliffside.lvl	2022-07-20 11:40:55.298394121 +0200
@@ -12,10 +12,10 @@
 #FALSE#
 #FALSE#
 "Cavi Cape.mp3"
[...]
```

5. And to add any Compat settings inbetween the comments and diff-content:

```diff
#1) Fill empty blocks for Multires.
#2) Add  bgo-61 (Black 32x32) to some pipes and blocks for improved looks.
#3) Set a compat value for example (not actually needed).
[compatibility]
enable-multipoints = false
--- OLD/Cliffside.lvl	2010-04-12 20:29:18.000000000 +0200
+++ NEW/Cliffside.lvl	2022-07-20 11:40:55.298394121 +0200
@@ -12,10 +12,10 @@
 #FALSE#
 #FALSE#
 "Cavi Cape.mp3"
[...]
```

Patch-Files will be stored inside patches.7z, which will be inside the ```settings``` folder.

This approach depends on:

- [ ] #132
- [ ] Implementing the MD5-System, see #54 
- [ ] The big Settings/Compat Rewrite that @ds-sloth and I have been talking about.

What do you think?

**Comment by 0lhi** [comment]
Since PhysFS can treat all contents of the Archive as if it were in a single folder, it may be better to organize stuff like this:

```
/Super RMN Bros.
-> /Cliffside.lvl
---->33e39e089734fa2b679898c40b5d6d77.patch
---->33e39e089734fa2b679898c40b5d6d77.compat.ini
```

**Comment by 0lhi** [comment]
The idea of a Patch System is officially **abandoned**.

Reasons for this decision are:

* The complication that it would add to the Engine
* The time it would take to maintain such a system
* The changes it would entail to the effort of preserving SMBX Episodes

We cancelled it in favor a different approach that is currently being worked on.

Additionally, we'll focus on repacking good Episodes so that they can take full advantage of TheXTech's new features.

**Reply by LoveBodhi** [reply]
That's a pain...

**Reply by ds-sloth** [reply]
From my perspective, maintaining this sort of database seems nearly impossible. Additionally, the technical issues with supporting this, especially on our lower-end targets, are quite great.

I agree that the efforts of interested community members are best spent on ensuring that episodes work well with modern settings.
